Doc/c-api/arg.rst

```diff
@@ -34,24 +34,39 @@ These formats allow accessing an object as a contiguous chunk of memory.
 You don't have to provide raw storage for the returned unicode or bytes
 area.
 
-In general, when a format sets a pointer to a buffer, the buffer is
-managed by the corresponding Python object, and the buffer shares
-the lifetime of this object.  You won't have to release any memory yourself.
-The only exceptions are ``es``, ``es#``, ``et`` and ``et#``.
-
-However, when a :c:type:`Py_buffer` structure gets filled, the underlying
-buffer is locked so that the caller can subsequently use the buffer even
-inside a :c:type:`Py_BEGIN_ALLOW_THREADS` block without the risk of mutable data
-being resized or destroyed.  As a result, **you have to call**
-:c:func:`PyBuffer_Release` after you have finished processing the data (or
-in any early abort case).
-
 Unless otherwise stated, buffers are not NUL-terminated.
 
-Some formats require a read-only :term:`bytes-like object`, and set a
-pointer instead of a buffer structure.  They work by checking that
-the object's :c:member:`PyBufferProcs.bf_releasebuffer` field is ``NULL``,
-which disallows mutable objects such as :class:`bytearray`.
+There are three ways strings and buffers can be converted to C:
+
+*  Formats such as ``y*`` and ``s*`` fill a :c:type:`Py_buffer` structure.
+   This locks the underlying buffer so that the caller can subsequently use
+   the buffer even inside a :c:type:`Py_BEGIN_ALLOW_THREADS`
+   block without the risk of mutable data being resized or destroyed.
+   As a result, **you have to call** :c:func:`PyBuffer_Release` after you have
+   finished processing the data (or in any early abort case).
+
+*  The ``es``, ``es#``, ``et`` and ``et#`` formats allocate the result buffer.
+   **You have to call** :c:func:`PyMem_Free` after you have finished
+   processing the data (or in any early abort case).
+
+*  .. _c-arg-borrowed-buffer:
+
+   Other formats take a read-only :term:`bytes-like object`, such as
+   :class:`bytes` or :class:`str`, and provide a ``const char *`` pointer to
```

Comment:
Are you sure about `str`? The `y` description below explicitly states "it does not accept Unicode objects", and it doesn't make sense to me that a buffer code would accept `str`.

Reply:
But `s` and `s#` accept `str`.

I think there is a wrong punctuation here. I would write either
```a read-only :term:`bytes-like object`, such as :class:`bytes`, or :class:`str` ```
(note the second comma)
or
```a read-only :term:`bytes-like object` (such as :class:`bytes`) or :class:`str` ```

But I do not know what is correct and looks better in English.

Or maybe change the order to avoid confusion?
```:class:`str` or a read-only :term:`bytes-like object`, such as :class:`bytes` ```

Reply:
Ah, that makes sense. I think putting `str` first (your last suggestion) would make this more clear.

Reply:
Thanks for the catch!